### ISSUE TITLE
Two identical functions behave differently in different files (and depending on their order)

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#6511]
**Describe the bug**
Two functions named `used` are type-checked differently despite being used identically in two files, the only difference being two other `_unused` functions which have different signatures but the same body as `used`.

**Code or Screenshots**

`bug1.py`:
```python
import numpy as np
import numpy.typing as npt

def _unused(foo: list[list[str]]) -> None:
    print(np.char.array(foo) + "X")

def used(foo: npt.NDArray[np.str_] | list[list[str]]) -> None:
    print(np.char.array(foo) + "X")

used([["A", "B"]])
```

`bug2.py`:
```python
import numpy as np
import numpy.typing as npt

def _unused(foo: list[str]) -> None:
    print(np.char.array(foo) + "X")

def used(foo: npt.NDArray[np.str_] | list[list[str]]) -> None:
    print(np.char.array(foo) + "X")

used([["A", "B"]])
```

Output:
```
(.venv) C:\Build\project>pyright bug1.py bug2.py 
C:\Build\project\bug1.py
  C:\Build\project\bug1.py:5:5 - error: Function "_unused" is not accessed (reportUnusedFunction)
C:\Build\project\bug2.py
  C:\Build\project\bug2.py:10:11 - error: Operator "+" not supported for types "_CharArray[bytes_]" and "Literal['X']" (reportGeneralTypeIssues)
  C:\Build\project\bug2.py:5:5 - error: Function "_unused" is not accessed (reportUnusedFunction)
3 errors, 0 warnings, 0 informations
```

The order matters:
```
(.venv) C:\Build\project>pyright bug2.py bug1.py 
C:\Build\project\bug2.py
  C:\Build\project\bug2.py:10:11 - error: Operator "+" not supported for types "_CharArray[bytes_]" and "Literal['X']" (reportGeneralTypeIssues)
  C:\Build\project\bug2.py:5:5 - error: Function "_unused" is not accessed (reportUnusedFunction)
C:\Build\project\bug1.py
  C:\Build\project\bug1.py:6:11 - error: Operator "+" not supported for types "_CharArray[bytes_]" and "Literal['X']" (reportGeneralTypeIssues)
  C:\Build\project\bug1.py:10:11 - error: Operator "+" not supported for types "_CharArray[bytes_]" and "Literal['X']" (reportGeneralTypeIssues)
  C:\Build\project\bug1.py:5:5 - error: Function "_unused" is not accessed (reportUnusedFunction)
5 errors, 0 warnings, 0 informations
```

**VS Code extension or command-line**
`numpy==1.26.2`, `pyright==1.1.336` via pyright-python.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

The code involves the use of `NDArray` which is a recursive type alias. Pyright uses an internal recursion limit to prevent stack overflows and combinatoric explosion (which can appear as a hang). We were hitting this limit in this example, which resulted in the strange behavior you observed. I've increased the recursion limit accordingly.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
